product_contract/models/sale_order.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2017 LasLabs Inc.
+# Copyright 2018 ACSONE SA/NV.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    is_contract = fields.Boolean(string="Is a contract", compute="_compute_is_contract")
+    contract_count = fields.Integer(compute="_compute_contract_count")
+    need_contract_creation = fields.Boolean(compute="_compute_need_contract_creation")
+
+    @api.constrains("state")
+    def check_contact_is_not_terminated(self):
+        for rec in self:
+            if rec.state not in (
+                "sale",
+                "done",
+                "cancel",
+            ) and rec.order_line.filtered("contract_id.is_terminated"):
+                raise ValidationError(
+                    _("You can't upsell or downsell a terminated contract")
+                )
+
+    @api.depends("order_line.contract_id", "state")
+    def _compute_need_contract_creation(self):
+        for rec in self:
+            rec.need_contract_creation = False
```

Comment:
```suggestion
            self.update({"need_contract_creation": False})
```
 is better for the call stack before the loop.

But, IMHO, you should build a set of ids for False values and update at the end only those ones.